token.go

```diff
@@ -0,0 +1,280 @@
+		hashFunc = crypto.SHA1
+		expectedCertHash = signingCertificate.Certificates[0].CertHash
+	} else {
+		// Reference: https://datatracker.ietf.org/doc/html/rfc5035#section-4
+		hashFunc = crypto.SHA256 // default hash algorithm for signingCertificateV2 is id-sha256
+		var ok bool
+		if signingCertificateV2.Certificates[0].HashAlgorithm.Algorithm != nil {
+			// use hash algorithm from SigningCertificateV2 signed attribute
+			hashFunc, ok = oid.ToHash(signingCertificateV2.Certificates[0].HashAlgorithm.Algorithm)
+			if !ok {
+				return nil, errors.New("unsupported certificate hash algorithm in SigningCertificateV2 attribute")
+			}
+		}
+		expectedCertHash = signingCertificateV2.Certificates[0].CertHash
+	}
+	certHash, err := hashutil.ComputeHash(hashFunc, candidateSigningCert.Raw)
+	if err != nil {
+		return nil, err
+	}
+	if !bytes.Equal(certHash, expectedCertHash) {
+		return nil, errors.New("signing certificate hash does not match CertHash in signed attribute")
+	}
+	return candidateSigningCert, nil
+}
+
+// Info returns the TSTInfo as defined in RFC 3161 2.4.2.
+//
+// Caller should use TSTInfo.Timestamp for consumption.
+func (t *SignedToken) Info() (*TSTInfo, error) {
+	var info TSTInfo
+	if _, err := asn1.Unmarshal(t.Content, &info); err != nil {
+		return nil, fmt.Errorf("cannot unmarshal TSTInfo from timestamp token: %w", err)
+	}
+	return &info, nil
+}
+
+//	Accuracy ::= SEQUENCE {
+//	 seconds     INTEGER             OPTIONAL,
+//	 millis  [0] INTEGER (1..999)    OPTIONAL,
+//	 micros  [1] INTEGER (1..999)    OPTIONAL }
+type Accuracy struct {
+	Seconds      int `asn1:"optional"`
+	Milliseconds int `asn1:"optional,tag:0"`
+	Microseconds int `asn1:"optional,tag:1"`
+}
+
+//	TSTInfo ::= SEQUENCE {
+//	 version         INTEGER                 { v1(1) },
+//	 policy          TSAPolicyId,
+//	 messageImprint  MessageImprint,
+//	 serialNumber    INTEGER,
+//	 genTime         GeneralizedTime,
+//	 accuracy        Accuracy                OPTIONAL,
+//	 ordering        BOOLEAN                 DEFAULT FALSE,
+//	 nonce           INTEGER                 OPTIONAL,
+//	 tsa             [0] GeneralName         OPTIONAL,
+//	 extensions      [1] IMPLICIT Extensions OPTIONAL }
+type TSTInfo struct {
+	Version        int // fixed to 1 as defined in RFC 3161 2.4.2
+	Policy         asn1.ObjectIdentifier
+	MessageImprint MessageImprint
+	SerialNumber   *big.Int
+	GenTime        time.Time        `asn1:"generalized"`
+	Accuracy       Accuracy         `asn1:"optional"`
+	Ordering       bool             `asn1:"optional,default:false"`
+	Nonce          *big.Int         `asn1:"optional"`
+	TSA            asn1.RawValue    `asn1:"optional,tag:0"`
+	Extensions     []pkix.Extension `asn1:"optional,tag:1"`
+}
+
+// Timestamp returns the timestamp by TSA and its accuracy.
+// tst MUST be valid and the time stamped datum MUST match message.
+func (tst *TSTInfo) Timestamp(message []byte) (time.Time, time.Duration, error) {
```

Comment:
This method name is kind of misleading. Instead of `Timestamp`, should we just call it `Verify` or `Validate`?

Reply:
The main purpose of this function is to extract the `GenTime` from the timestamp token's `TSTInfo`. How about `ExtractGenTime`?